File: Java-Implementation/TaskManager.java
package todoapp;

import java.io.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class TaskManager {
    private List<Task> tasks;
    private List<User> users;
    private List<Category> categories;
    private final ReadWriteLock lock = new ReentrantReadWriteLock();
    private static final String DATA_FILE = "tasks.dat";
    private static final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd");

    public TaskManager() {
        tasks = new ArrayList<>();
        users = new ArrayList<>();
        categories = new ArrayList<>();
        loadData();
    }

    public void addUser(String username) {
        lock.writeLock().lock();
        try {
            for (User user : users) {
                if (user.getUsername().equals(username)) {
                    System.out.println("User already exists!");
                    return;
                }
            }
            users.add(new User(username));
            System.out.println("User added successfully!");
        } finally {
            lock.writeLock().unlock();
        }
    }

    public User findUser(String username) {
        lock.readLock().lock();
        try {
            for (User user : users) {
                if (user.getUsername().equals(username)) {
                    return user;
                }
            }
            return null;
        } finally {
            lock.readLock().unlock();
        }
    }

    public void addCategory(String name) {
        lock.writeLock().lock();
        try {
            for (Category category : categories) {
                if (category.getName().equals(name)) {
                    System.out.println("Category already exists!");
                    return;
                }
            }
            categories.add(new Category(name));
            System.out.println("Category added successfully!");
        } finally {
            lock.writeLock().unlock();
        }
    }

    public Category findCategory(String name) {
        lock.readLock().lock();
        try {
            for (Category category : categories) {
                if (category.getName().equals(name)) {
                    return category;
                }
            }
            return null;
        } finally {
            lock.readLock().unlock();
        }
    }

    public void addTask(String title, String description, String dueDateStr, String categoryName, String assignedUsername) {
        lock.writeLock().lock();
        try {
            Date dueDate = null;
            try {
                dueDate = dateFormat.parse(dueDateStr);
            } catch (ParseException e) {
                System.out.println("Invalid date format. Use yyyy-MM-dd");
                return;
            }

            Category category = findCategory(categoryName);
            if (category == null) {
                System.out.println("Category not found!");
                return;
            }

            Task task = new Task(title, description, dueDate, category);
            if (assignedUsername != null && !assignedUsername.isEmpty()) {
                User user = findUser(assignedUsername);
                if (user == null) {
                    System.out.println("User not found!");
                    return;
                }
                task.setAssignedUser(user);
                user.assignTask(task);
            }

            tasks.add(task);
            category.addTask(task);
            System.out.println("Task added successfully!");
            saveData();
        } finally {
            lock.writeLock().unlock();
        }
    }

    public void markTaskAsComplete(int taskId) {
        lock.writeLock().lock();
        try {
            Task task = findTaskById(taskId);
            if (task != null) {
                task.setCompleted(true);
                System.out.println("Task marked as complete!");
                saveData();
            } else {
                System.out.println("Task not found!");
            }
        } finally {
            lock.writeLock().unlock();
        }
    }

    public void deleteTask(int taskId) {
        lock.writeLock().lock();
        try {
            Task task = findTaskById(taskId);
            if (task != null) {
                tasks.remove(task);
                task.getCategory().removeTask(task);
                if (task.getAssignedUser() != null) {
                    task.getAssignedUser().removeTask(task);
                }
                System.out.println("Task deleted successfully!");
                saveData();
            } else {
                System.out.println("Task not found!");
            }
        } finally {
            lock.writeLock().unlock();
        }
    }

    public Task findTaskById(int taskId) {
        for (Task task : tasks) {
            if (task.getId() == taskId) {
                return task;
            }
        }
        return null;
    }

    public List<Task> getAllTasks() {
        lock.readLock().lock();
        try {
            return new ArrayList<>(tasks);
        } finally {
            lock.readLock().unlock();
        }
    }

    public List<Task> getTasksByUser(String username) {
        lock.readLock().lock();
        try {
            User user = findUser(username);
            if (user != null) {
                return new ArrayList<>(user.getAssignedTasks());
            }
            return new ArrayList<>();
        } finally {
            lock.readLock().unlock();
        }
    }

    public List<Task> getTasksByCategory(String categoryName) {
        lock.readLock().lock();
        try {
            Category category = findCategory(categoryName);
            if (category != null) {
                return new ArrayList<>(category.getTasks());
            }
            return new ArrayList<>();
        } finally {
            lock.readLock().unlock();
        }
    }

    public List<User> getAllUsers() {
        lock.readLock().lock();
        try {
            return new ArrayList<>(users);
        } finally {
            lock.readLock().unlock();
        }
    }

    public List<Category> getAllCategories() {
        lock.readLock().lock();
        try {
            return new ArrayList<>(categories);
        } finally {
            lock.readLock().unlock();
        }
    }

    private void saveData() {
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(DATA_FILE))) {
            oos.writeObject(tasks);
            oos.writeObject(categories);
            System.out.println("Data saved successfully!");
        } catch (IOException e) {
            System.out.println("Error saving data: " + e.getMessage());
        }
    }

    @SuppressWarnings("unchecked")
    private void loadData() {
        File file = new File(DATA_FILE);
        if (!file.exists()) {
            System.out.println("No existing data found. Starting fresh.");
            initializeDefaultData();
            return;
        }

        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(DATA_FILE))) {
            tasks = (List<Task>) ois.readObject();
            categories = (List<Category>) ois.readObject();
            
            // Reconstruct users from tasks
            Set<String> usernames = new HashSet<>();
            for (Task task : tasks) {
                if (task.getAssignedUser() != null) {
                    usernames.add(task.getAssignedUser().getUsername());
                }
            }
            
            for (String username : usernames) {
                users.add(new User(username));
            }
            
            System.out.println("Data loaded successfully!");
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Error loading data: " + e.getMessage());
            initializeDefaultData();
        }
    }
    
    private void initializeDefaultData() {
        // Add default categories
        categories.add(new Category("Work"));
        categories.add(new Category("Personal"));
        categories.add(new Category("Study"));
        
        // Add default users
        users.add(new User("admin"));
        users.add(new User("user1"));
        
        System.out.println("Default data initialized.");
    }
}

File: ToDoListNodeJs/taskManager.js
// taskManager.js
const fs = require('fs').promises;
const path = require('path');
const Task = require('./task');
const EventEmitter = require('events');

const DATA_FILE = path.join(__dirname, 'tasks.json');

class TaskManager extends EventEmitter {
    constructor() {
        super();
        this.tasks = [];
        this.users = [];
        this.categories = [];
        this.nextTaskId = 1;
    }

    async init() {
        try {
            await this.loadData();
            console.log('Data loaded successfully!');
        } catch (error) {
            console.log('No existing data found or error loading data. Starting fresh.');
            this.initializeDefaultData();
        }
    }

    initializeDefaultData() {
        // Add default categories
        this.categories = ['Work', 'Personal', 'Study'];
        
        // Add default users
        this.users = ['admin', 'user1'];
        
        this.tasks = [];
        this.nextTaskId = 1;
        
        console.log('Default data initialized.');
    }

    async loadData() {
        const data = await fs.readFile(DATA_FILE, 'utf8');
        const parsedData = JSON.parse(data);
        
        this.tasks = parsedData.tasks;
        this.users = parsedData.users;
        this.categories = parsedData.categories;
        this.nextTaskId = parsedData.nextTaskId;
    }

    async saveData() {
        const data = {
            tasks: this.tasks,
            users: this.users,
            categories: this.categories,
            nextTaskId: this.nextTaskId
        };
        
        await fs.writeFile(DATA_FILE, JSON.stringify(data, null, 2), 'utf8');
        console.log('Data saved successfully!');
    }

    async addUser(username) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                if (this.users.includes(username)) {
                    reject(new Error('User already exists!'));
                    return;
                }
                
                this.users.push(username);
                this.saveData().then(() => {
                    this.emit('userAdded', username);
                    resolve(username);
                }).catch(reject);
            }, 100); // Simulate some async operation
        });
    }

    async addCategory(name) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                if (this.categories.includes(name)) {
                    reject(new Error('Category already exists!'));
                    return;
                }
                
                this.categories.push(name);
                this.saveData().then(() => {
                    this.emit('categoryAdded', name);
                    resolve(name);
                }).catch(reject);
            }, 100); // Simulate some async operation
        });
    }

    async addTask(title, description, dueDate, category, assignedUser = null) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                if (!this.categories.includes(category)) {
                    reject(new Error('Category not found!'));
                    return;
                }
                
                if (assignedUser && !this.users.includes(assignedUser)) {
                    reject(new Error('User not found!'));
                    return;
                }
                
                const task = {
                    id: this.nextTaskId++,
                    title,
                    description,
                    completed: false,
                    dueDate,
                    category,
                    assignedUser
                };
                
                this.tasks.push(task);
                this.saveData().then(() => {
                    this.emit('taskAdded', task);
                    resolve(task);
                }).catch(reject);
            }, 200); // Simulate some async operation
        });
    }

    async markTaskAsComplete(taskId) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                const task = this.tasks.find(t => t.id === taskId);
                if (!task) {
                    reject(new Error('Task not found!'));
                    return;
                }
                
                task.completed = true;
                this.saveData().then(() => {
                    this.emit('taskCompleted', task);
                    resolve(task);
                }).catch(reject);
            }, 150); // Simulate some async operation
        });
    }

    async deleteTask(taskId) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                const taskIndex = this.tasks.findIndex(t => t.id === taskId);
                if (taskIndex === -1) {
                    reject(new Error('Task not found!'));
                    return;
                }
                
                const task = this.tasks[taskIndex];
                this.tasks.splice(taskIndex, 1);
                this.saveData().then(() => {
                    this.emit('taskDeleted', task);
                    resolve(task);
                }).catch(reject);
            }, 150); // Simulate some async operation
        });
    }

    async getAllTasks() {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve([...this.tasks]);
            }, 100);
        });
    }

    async getTasksByUser(username) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                if (!this.users.includes(username)) {
                    reject(new Error('User not found!'));
                    return;
                }
                
                const userTasks = this.tasks.filter(task => task.assignedUser === username);
                resolve(userTasks);
            }, 100);
        });
    }

    async getTasksByCategory(categoryName) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                if (!this.categories.includes(categoryName)) {
                    reject(new Error('Category not found!'));
                    return;
                }
                
                const categoryTasks = this.tasks.filter(task => task.category === categoryName);
                resolve(categoryTasks);
            }, 100);
        });
    }

    async getAllUsers() {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve([...this.users]);
            }, 50);
        });
    }

    async getAllCategories() {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve([...this.categories]);
            }, 50);
        });
    }
}

module.exports = TaskManager;